### PR TITLE
feat: merge per-worker flow entries in frontend stats client

### LINKED ISSUE #156: Per-worker counter tables with merge-at-snapshot to eliminate cross-core CAS
## Problem

Currently all VPP worker threads share a single `counter_table` per flow rule (in `infmon_plugin_main_t.tables[flow_rule_index]`). Workers use atomic CAS to insert/update slots concurrently. This design:

- Causes cross-core cache-line bouncing on every counter update (the hot path).
- Makes `snapshot_and_clear` swap a single pointer, but workers on different cores may still be mid-CAS on the old table when the swap happens.
- Spec 004 §12.3 notes this as an open question ("Per-CPU sharded tables") deferred to v2.

With RSS distributing packets across workers, each worker thread processes its own RX queue independently. Giving each worker its own table eliminates all cross-core atomics on the data path.

### Backend changes

1. Change `infmon_plugin_main_t.tables` from `[INFMON_MAX_ACTIVE_FLOW_RULES]` to `[INFMON_MAX_WORKERS][INFMON_MAX_ACTIVE_FLOW_RULES]`.
2. `infmon_counter_update` receives `tables[worker_id]` instead of the shared array.
3. `infmon_snapshot_and_clear` swaps ALL per-worker tables for a given flow_rule_index atomically (one RELEASE store per worker), then enqueues all old tables for RCU retirement.
4. The VAPI reply for `snapshot_inline_dump` streams slots from all per-worker retired tables.

### Frontend changes

5. The poller receives slots from multiple per-worker tables and merges them: same key across workers → sum packets/bytes, take max(last_update).

### Spec changes

6. Update spec 004 §5 to describe per-worker table layout.
7. Update spec 004 §7.2 to describe per-worker swap semantics.
8. Update spec 004 §12.3 to mark this as resolved (no longer deferred).
9. Update spec 005 §8.1 to describe per-worker merge in the frontend.

## Performance impact

- Eliminates all cross-core CAS on the data path (the #1 bottleneck).
- Memory cost: N_workers × N_flow_rules × table_size (acceptable on BF-3 with GiBs of hugepage).
- Snapshot latency increases slightly (swap N pointers instead of 1), but still well within the 5 ms target.

### PR DESCRIPTION
**Pull request by r12f**
## Summary

Implements the frontend merge logic for per-worker counter tables (#156).

### Problem

With per-worker counter tables, the backend's `snapshot_inline_dump` handler streams slots from **all** worker tables independently. When multiple VPP workers process packets for the same flow, the frontend receives duplicate entries — one per worker — with partial counters.

### Solution

Add `merge_worker_entries()` in the VAPI stats client that groups entries by `(key_hash, key_bytes)` identity and merges:
- **packets/bytes**: saturating sum across workers
- **last_update**: take the maximum

This runs before building the `RawDescriptor`, so the existing decode pipeline sees already-merged data with no changes needed downstream.

### Also included

Fix VAPI FFI build break: `vapi_infmon_flow_rule_{add,del}` signatures changed after API regeneration — now require callback+context args. Pass `NULL, NULL` for blocking mode.

### Tests

- 7 new unit tests for `merge_worker_entries`:
  - No duplicates (passthrough)
  - Two/three workers same key
  - Mixed unique + duplicate keys
  - Hash collision (different keys, same hash — not merged)
  - Empty input
  - Saturating arithmetic at u64::MAX
- All 85 existing tests pass

Closes #156

### CODE REVIEW

File: src/frontend/src/vapi_stats_client.rs [hunk 1]
```diff
@@ -143,16 +143,17 @@ impl VapiStatsClient {
                 continue;
             }
 
-            // Build a RawDescriptor from the collected entries.
-            // We assemble a key_arena by concatenating all key bytes.
+            // Merge entries that share the same flow key across workers.
+            // With per-worker counter tables, the same key can appear once per
+            // worker. We merge: sum packets/bytes, max last_update.
             let first = &entries[0];
```
Comment (banidoru) [inline src/frontend/src/vapi_stats_client.rs:153]: Nit: `first` is still bound to `entries[0]` (pre-merge) and only used later for `RawDescriptor` metadata (`generation`, `epoch_ns`, `insert_failed`, `table_full`). This works because those fields are snapshot-level and identical across entries, but it's non-obvious post-refactor.

Consider a brief comment, or pull metadata from `merged[0]` to keep the data lineage clearer.

File: src/frontend/src/vapi_stats_client.rs [hunk 2]
```diff
@@ -193,6 +194,48 @@ impl Drop for VapiStatsClient {
...
 
+/// Merge entries from multiple workers that share the same flow key.
+///
+/// Identity: `(key_hash, key_bytes)`. For matching entries we sum
+/// `packets` and `bytes`, and take `max(last_update)`.
+fn merge_worker_entries(entries: &[CollectedEntry]) -> Vec<CollectedEntry> {
+    use std::collections::HashMap;
+
+    // Map from (key_hash, key_bytes) → index into `merged`.
+    let mut index_map: HashMap<(u64, &[u8]), usize> = HashMap::new();
```
Comment (banidoru) [inline src/frontend/src/vapi_stats_client.rs]: Minor perf: for large flow tables, pre-allocating avoids rehashing:
```rust
let mut index_map = HashMap::with_capacity(entries.len());
let mut merged = Vec::with_capacity(entries.len());
```
`entries.len()` is an upper bound (exact when no duplicates).

File: src/frontend/src/vapi_stats_client.rs [hunk 3]
```diff
@@ -193,6 +194,48 @@ impl Drop for VapiStatsClient {
...
+            merged.push(CollectedEntry {
+                _flow_rule_id_hi: e._flow_rule_id_hi,
+                _flow_rule_id_lo: e._flow_rule_id_lo,
+                generation: e.generation,
+                epoch_ns: e.epoch_ns,
+                insert_failed: e.insert_failed,
+                table_full: e.table_full,
+                key_hash: e.key_hash,
+                packets: e.packets,
+                bytes: e.bytes,
```
Comment (banidoru) [inline src/frontend/src/vapi_stats_client.rs:243]: `insert_failed` and `table_full` are copied verbatim from the first worker that creates a merge slot. With per-worker counter tables, each worker tracks its own insert-failure / table-full count independently. The merge should aggregate these (e.g. saturating sum), otherwise we silently drop failure counts from workers 1..N.

Same applies to `generation` and `epoch_ns` if they can differ across workers — worth at least a debug-assert that they're equal.

**Review by banidoru — COMMENTED** [review]
- Clean merge logic, good use of `(key_hash, key_bytes)` composite identity and saturating arithmetic
- `insert_failed` / `table_full` only taken from first worker entry — should be aggregated across workers
- `first` still references pre-merge `entries[0]` for `RawDescriptor` metadata — works but fragile after refactor
- Minor: `HashMap`/`Vec` could use `with_capacity` to avoid rehashing on large tables
- Tests are thorough — hash collision, saturation, mixed cases all covered

File: src/frontend/src/vapi_stats_client.rs [hunk 4]
```diff
@@ -193,6 +194,48 @@ impl Drop for VapiStatsClient {
...
+                m.last_update = e.last_update;
+            }
+        } else {
+            let idx = merged.len();
+            index_map.insert(identity, idx);
+            merged.push(CollectedEntry {
+                _flow_rule_id_hi: e._flow_rule_id_hi,
+                _flow_rule_id_lo: e._flow_rule_id_lo,
+                generation: e.generation,
+                epoch_ns: e.epoch_ns,
```
Comment (banidoru) [inline src/frontend/src/vapi_stats_client.rs:238]: `generation` and `epoch_ns` are copied from the first worker that creates a merge slot (same concern as `insert_failed`/`table_full` raised elsewhere). If workers snapshot at slightly different times, `epoch_ns` could diverge. Consider taking `max(epoch_ns)` or documenting that all workers share identical snapshot-level metadata.

File: src/frontend/src/vapi_ffi/infmon_vapi_client.c [hunk 5]
```diff
@@ -252,7 +252,7 @@ int infmon_vapi_flow_rule_add(void *handle, const char *name, const uint8_t *fie
     msg->payload.eviction_policy = eviction_policy;
 
     /* Send as blocking request */
-    vapi_error_e rv = vapi_infmon_flow_rule_add(ctx, msg);
+    vapi_error_e rv = vapi_infmon_flow_rule_add(ctx, msg, NULL, NULL);
```
Comment (banidoru) [inline src/frontend/src/vapi_ffi/infmon_vapi_client.c:255]: The added `NULL, NULL` arguments suggest an upstream API signature change. This is orthogonal to the per-worker merge feature — worth a note in the PR description or a separate commit so the merge change is self-contained.

**Review by banidoru — COMMENTED** [review]
- Merge logic is correct and well-tested; good use of saturating arithmetic and `(key_hash, key_bytes)` composite identity
- `generation` and `epoch_ns` are silently taken from the first worker entry in the merge — if workers can diverge (e.g. different snapshot timing), this should be documented or explicitly reconciled
- The C FFI signature change (`NULL, NULL` args) appears unrelated to the per-worker merge feature — consider splitting into a separate commit or noting in the PR description
- Test coverage is thorough, including hash-collision and overflow edge cases

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

File: src/frontend/src/vapi_stats_client.rs (hunk 1 above)
Comment (r12f) [reply src/frontend/src/vapi_stats_client.rs:153]: Good call — added a clarifying comment explaining that `first` is used only for snapshot-level metadata (`generation`, `epoch_ns`) which is identical across workers for the same flow rule. Fixed in 88ce5d9.

File: src/frontend/src/vapi_stats_client.rs (hunk 2 above)
Comment (r12f) [reply src/frontend/src/vapi_stats_client.rs]: Done — pre-allocated both `HashMap::with_capacity(entries.len())` and `Vec::with_capacity(entries.len())`. `entries.len()` is an upper bound (exact when no duplicates). Fixed in 88ce5d9.

File: src/frontend/src/vapi_stats_client.rs (hunk 3 above)
Comment (r12f) [reply src/frontend/src/vapi_stats_client.rs:243]: Excellent catch — `insert_failed` and `table_full` are now aggregated via `saturating_add` during the merge (both in `merge_worker_entries` for per-entry merging and in the `RawDescriptor` build which now sums across all merged entries instead of taking only `first`). This ensures failure counts from all workers are preserved. Fixed in 88ce5d9.

File: src/frontend/src/vapi_stats_client.rs (hunk 4 above)
Comment (banidoru) [reply src/frontend/src/vapi_stats_client.rs:238]: Addressed — the added comment clearly documents the assumption that `generation` and `epoch_ns` are identical across workers. Thanks.

File: src/frontend/src/vapi_ffi/infmon_vapi_client.c (hunk 5 above)
Comment (banidoru) [reply src/frontend/src/vapi_ffi/infmon_vapi_client.c:255]: Still open — the `NULL, NULL` change is still bundled without a note in the PR description or a separate commit.

File: src/frontend/src/vapi_stats_client.rs (hunk 4 above)
Comment (r12f) [reply src/frontend/src/vapi_stats_client.rs:238]: Good point — added `max(epoch_ns)` during merge so we always keep the latest snapshot timestamp. `generation` is set at rule-add time and is identical across workers, so no merge needed — added a comment documenting this. Fixed in 856a987.

File: src/frontend/src/vapi_stats_client.rs [hunk 6]
```diff
@@ -173,8 +178,14 @@ impl VapiStatsClient {
...
                 key_arena,
-                insert_failed: first.insert_failed,
-                table_full: first.table_full,
+                insert_failed: merged
+                    .iter()
+                    .map(|e| e.insert_failed)
+                    .fold(0u64, u64::saturating_add),
+                table_full: merged
+                    .iter()
+                    .map(|e| e.table_full)
```
Comment (banidoru) [inline src/frontend/src/vapi_stats_client.rs]: `insert_failed` and `table_full` are now summed in two places:
1. Inside `merge_worker_entries` — per-key accumulation via `saturating_add` (line 225)
2. Here — summing across all merged entries in the descriptor build

Since these are **table-level** counters (not per-flow), every entry from the same worker carries the same value. After merge, each merged entry already has the sum of its workers' values. Then summing across all *keys* here massively over-counts.

Pick one aggregation site. Simplest fix: remove the per-entry accumulation from `merge_worker_entries` (keep `insert_failed`/`table_full` from the first entry only), and aggregate here across **unique workers** — or just take `first.insert_failed`/`first.table_full` if these are truly table-level snapshots that are identical across workers (as stated in the comment for `generation`/`epoch_ns`).

**Review by banidoru — COMMENTED** [review]
Re-review (iteration 2):

- Prior threads on metadata comment, pre-allocation, and `insert_failed`/`table_full` aggregation all addressed — nice.
- `epoch_ns`/`generation` concern resolved with clarifying comment.
- **New issue:** `insert_failed`/`table_full` double-counted — summed per-key in `merge_worker_entries` AND again across all merged entries in descriptor build. Pick one aggregation site.
- `NULL, NULL` FFI change still bundled without explanation — consider a PR description note or separate commit.

File: src/frontend/src/vapi_ffi/infmon_vapi_client.c (hunk 5 above)
Comment (r12f) [reply src/frontend/src/vapi_ffi/infmon_vapi_client.c:255]: Acknowledged — the `NULL, NULL` arguments are due to an upstream VAPI signature change (added callback + context params for async support). This was a pre-existing call-site update that landed on this branch. Adding a note in the PR description for clarity.

File: src/frontend/src/vapi_stats_client.rs (hunk 6 above)
Comment (r12f) [reply src/frontend/src/vapi_stats_client.rs]: Great catch — you're right, this was double-counting. Since insert_failed/table_full are table-level snapshot counters (identical across all entries from the same worker), removed the per-key saturating_add from merge_worker_entries and changed the descriptor build to use first.insert_failed/first.table_full directly. Fixed in 19f64e0.

Comment (banidoru) [reply src/frontend/src/vapi_stats_client.rs]: Double-counting fix in 19f64e0 looks correct — removed both the per-key accumulation in `merge_worker_entries` and the cross-key sum in the descriptor build.

One remaining subtlety: `first.insert_failed` / `first.table_full` now captures only **one worker's** table-level counters. With per-worker counter tables, each worker tracks failures independently, so this under-reports if multiple workers hit insert failures. A correct aggregate would deduplicate by worker (e.g. track seen worker IDs) and sum across unique workers only. Not a blocker for this PR — the previous double-counting was strictly worse — but worth a follow-up.

**Review by banidoru — APPROVED** [review]
- Double-counting of `insert_failed`/`table_full` fixed correctly in 19f64e0 — both per-key and cross-key aggregation removed
- `max(epoch_ns)` merge added per prior feedback — good
- Minor: `first.insert_failed` only captures one worker's failure count; cross-worker aggregation deferred as follow-up (acceptable)
- Test coverage is solid — hash collision, saturation, empty input all covered
- LGTM

**Review by banidoru — APPROVED** [review]
- All prior review threads addressed and resolved — double-counting fix, pre-allocation, epoch_ns max-merge, metadata comments all landed correctly
- Merge identity `(key_hash, key_bytes)` is sound; hash-collision test confirms no false merges
- Saturating arithmetic and 7 unit tests provide solid coverage for edge cases
- Minor follow-up opportunity: `first.insert_failed`/`table_full` captures only one worker's table-level counters (under-reports with multiple workers hitting failures) — noted by prior reviewer as non-blocker
- LGTM
